### ISSUE TITLE
Support for Low-Latency HLS (LL-HLS)

### ISSUE THREAD

**Issue by ezadoo** [AirenSoft/OvenMediaEngine#480]
**Is your feature request related to a problem? Please describe.**
As OME supports LL-Dash which is, as far as I researched, not working on iOS, also LL-HLS could be supported.

**Describe the solution you'd like**
As LL-HLS works with the same CMAF-Format, which is also already used by LL-Dash, wouldn't it be possible to generate an additional manifest linking to the same CMAF-Files?
So both, LL-HLS and LL-Dash could be supported, if the CMAF-Files are not encrypted.
 
**Describe alternatives you've considered**
Currently the only alternative for iOS is to not use low-latency streaming for iOS-Devices.

**Additional context**

**Comment by mpisat** [comment]
LL-HLS would be a killer feature. It'd be great!

**Comment by catapop84** [comment]
Indeed. LL-HLS will be a killer feature.

**Comment by getroot** [comment]
Is there a full reference player for playing LLHLS released? If you have any related information, please share. Until recently, I was seeing the draft of the LLHLS protocol being updated and unconfirmed.

**Comment by ezadoo** [comment]
What do you mean with reference player?

As for now all iOS Devices running iOS14 or higher are natively supporting LL-HLS, as far as I know.
Because LL-HLS was developed by Apple I would think that this could be counted as the reference implementation/player.
iOS 14 was released nearly a year ago, so there is a quite high market share among the iOS Devices.
(I call it simply iOS14, but this includes iPadOS14 and tvOS14.)

So it would be very useful for native iOS Apps, but Safari is supporting LL-HLS on these devices too with the native player.

Also there are already some third party player frameworks, which support LL-HLS with CMAF across multiple platforms.

There is to be mentioned theoplayer and also jwplayer.

The developers of hls.js are also working on the integration, and should be ready soon.

ShakaPlayer is also supporting LL-HLS, but doesn't support CMAF, also not with LL-Dash, if I understood correctly.

The VHS extension for videojs also states that [it should be all implemented, but it currently lacks testing](https://github.com/videojs/http-streaming/issues/721#issuecomment-890026893), so it isn't ready for production, but should in near future.

**Comment by nums** [comment]
@getroot, Apple released demo scripts in PHP and GO : https://developer.apple.com/download/all/?q=hls
If you don't have an apple developper account, I can share you the dmg file or convert it to zip

**Comment by basisbit** [comment]
hls.js supports LL-HLS. I'd suggest using that as reference player.
Anyways, before implementing such new stuff, please first fix the existing features and work on implementing the basics that will also be required for LL-HLS, like HTTP-Keep-Alife or HTTP 2 support. Without that, LL-HLS would fail in production just like LL-DASH currently does.

For those here who ask for LL-HLS support, may I ask what use case there is for LL-HLS, where WebRTC is not a much better solution?

**Comment by basisbit** [comment]
@ezadoo you wrote `Currently the only alternative for iOS is to not use low-latency streaming for iOS-Devices.`, however, we regularly used WebRTC based high quality video streaming using OvenMediaEngine just fine recently during events with 35k+ viewers.

**Comment by ezadoo** [comment]
@basisbit I'm going completely with your first comment, that the existing issues need to be addressed respectively must be addressed so that LL-HLS can work properly.

You're right, if I simply want low latency, WebRTC could do the job too, with all the negative consequences on the server side.

I don't want to start here a discussion whether using WebRTC or other protocols nor the pros and cons of WebRTC in general (Scalability, CDNs, Quality, etc).

So I keep my answer short:
LL-HLS has merely the same use case as LL-Dash but is more widely supported (as it's working on iOS-Devices, but the support of Dash on iOS Devices is quite bad), and LL-HLS is in most cases also supported too, where Dash is supported.
So LL-HLS would be the better choice over LL-Dash in my eyes.

But as both can use the same CMAF-Files we would "only" need an additional playlist linking to the same files, which are already created, and not a decision between LL-Dash and LL-HLS.

**Comment by basisbit** [comment]
TLDR: I disagree and actually have proven that with more than 20 different weekend-events in the last couple of months where each of them had something like 15k to 35k attendees watching over WebRTC from Asia, Oceania, Europe, North America and South America, while using something like 100$ of budget per event for hourly billed servers and traffic fees.

> So for multiple thousands of viewers WebRTC is a bad choice

This imho is not the case at all. Regarding video quality, you can send approximately the same h.264 or h.265 encoded frame content over WebRTC like over *HLS. The webbrowser will internally even use mostly the same code to play back the data, as both use a HTML5 video element tag and the difference is only how the playback buffer is filled, how much data is buffered and with what data the buffer is filled. Only on the encoding side, you'll need a bit more GPU/CPU resources to get approximately the same quality into the stream without having as much look-ahead frame data, and not being able to use b-frames.

Regarding Scalability / CDN: Just do it yourself. Much cheaper, less single point of failure, and you get low latency video streaming for free once you want that. LL-HLS doesn't scale well over the typical NginX reverse proxy CDNs either, unless you go for the expensive "CDN+" things. If you want to just have another company deal with distribution for you, set your HLS chunk size to 2 or 3 seconds, and use BunnyCDN with their volume tier, so you are at ~ 5$ per TB for reverse-proxied plus decently cached traffic. That will give you 6 or 9 seconds of server to user latency. With LL-HLS, you'll still be at something like 4 to 6 seconds of edge server to user latency when doing international streaming.

**Comment by getroot** [comment]
LLHLS released in pre-alpha. Please discuss further in the issues below. thank you!

https://github.com/AirenSoft/OvenMediaEngine/issues/766

**Comment by getroot** [comment]
Closing this issue now that LLHLS has been released. For further discussion of LLHLS, please visit https://github.com/AirenSoft/OvenMediaEngine/issues/766 or create a new issue.

Thank you for contribution!

**Comment by fastfading** [comment]
does it support llhls  fmp4 encryption ?
